File: Unity/Assets/Scripts/Building/Floorplan.cs
using System.Collections.Generic;

namespace Vacancy
{
    public static class AreaKind
    {
        public const string Corridor = "corridor";
        public const string Lobby = "lobby";
        public const string GuestRoom = "guestRoom";
        public const string Office = "office";
        public const string Department = "department";
        public const string Parking = "parking";
        public const string Stairs = "stairs";
        public const string Basement = "basement";
        public const string Storage = "storage";
        public const string Walkway = "walkway";
    }

    public sealed class BandSpec
    {
        public string Kind;
        public string DoorSide;
        public float Height;
        public bool Grow;
    }

    public sealed class FloorSpec
    {
        public string Id;
        public string Name;
        public int Level;
        public float Tile;
        public float Edge;
        public float SideCorridor;
        public float DoorWidth;
        public Point RoomSize;
        public int MaxRoomsPerRow;
        public List<BandSpec> Bands;
    }

    public sealed class FloorArea
    {
        public string Id;
        public string Kind;
        public string Token;
        public string Label;
        public string Accent;
        public Rect Rect;
        public bool Walls;
        public List<Door> Doors = new List<Door>();
        public int? RoomId;
        public string DepartmentId;
        public int Level;
    }

    public sealed class PlannedRoom
    {
        public int Id;
        public Rect Rect;
        public Point Center;
        public string DoorSide;
        public Point Door;
        public Door DoorOpening;
        public Point Approach;
        public int Level;
    }

    public sealed class DepartmentSpot
    {
        public string Id;
        public string Label;
        public string Accent;
        public Rect Rect;
        public float X;
        public float Y;
        public float W;
        public float H;
    }

    public sealed class OfficeSpot
    {
        public float X;
        public float Y;
        public float W;
        public float H;
        public Rect Rect;
        public Point Door;
        public Point Approach;
    }

    public sealed class DeskSpot
    {
        public float X;
        public float Y;
        public float W;
        public float H;
    }

    public sealed class BuiltFloor
    {
        public string Id;
        public string Name;
        public int Level;
        public float Tile;
        public FloorSpec Spec;
        public Rect Bounds;
        public Rect Content;
        public List<FloorArea> Areas = new List<FloorArea>();
        public List<PlannedRoom> Rooms = new List<PlannedRoom>();
        public int RoomsPerRow;
        public Point RoomSize;
        public Rect Lobby;
        public Rect Parking;
        public Rect Basement;
        public Rect Stairs;
        public Rect UpperStairs;
        public OfficeSpot Office;
        public DeskSpot FrontDesk;
        public Dictionary<string, DepartmentSpot> Departments = new Dictionary<string, DepartmentSpot>();
        public Rect WalkWest;
        public Rect WalkNorth;
        public Rect PorteCochere;
        public Rect DriveSouth;
        public Rect CornerMass;
        public Point LotEntrance;
        public Point LotEntranceEast;
        public Point LotEntranceSouth;
        public bool OutdoorCourt;
    }

    public static class Floorplan
    {
        public static readonly FloorSpec FlagshipGround = new FloorSpec
        {
            Id = "ground",
            Name = "Inn",
            Level = 0,
            Tile = 10,
            Edge = 20,
            SideCorridor = 100,
            DoorWidth = 40,
            RoomSize = new Point(GameConfig.RoomWidth, GameConfig.RoomHeight),
            MaxRoomsPerRow = 8,
            Bands = new List<BandSpec>()
        };

        static readonly HashSet<string> PublicKinds = new HashSet<string>
        {
            AreaKind.Corridor, AreaKind.Lobby, AreaKind.Department, AreaKind.Parking,
            AreaKind.Stairs, AreaKind.Basement, AreaKind.Storage, AreaKind.Walkway
        };

        static readonly DepartmentDef[] Departments =
        {
            new DepartmentDef("housekeeping", "Mary's room", "left", "#e8a0bf"),
            new DepartmentDef("maintenance", "Bob's closet", "right", "#ffb347")
        };

        struct DepartmentDef
        {
            public readonly string Id;
            public readonly string Label;
            public readonly string Side;
            public readonly string Accent;

            public DepartmentDef(string id, string label, string side, string accent)
            {
                Id = id;
                Label = label;
                Side = side;
                Accent = accent;
            }
        }

        public static BuiltFloor CreateMotorCourt(FloorSpec spec = null)
        {
            spec = spec ?? FlagshipGround;
            float tile = spec.Tile;
            float Down(float value) => (float)(System.Math.Floor(value / tile) * tile);

            // Facade along the courtyard walk; depth back from the door.
            float roomAlong = Down(spec.RoomSize.X);
            float roomDepth = Down(spec.RoomSize.Y);
            float walkW = Down(spec.SideCorridor > 0 ? spec.SideCorridor : 100f);
            float originX = Down(40);
            float originY = Down(40);
            const int westCount = 8;
            const int northCount = 4;
            // Leftover bays beside the office were ~30 units; grow the lobby
            // into unused dirt (west) and the drive (east) so both sides fill.
            float lobbySide = Down(40);
            float lobbyW = Down(280) + lobbySide * 2f;
            float lobbyH = Down(280);
            float canopyH = Down(80);

            var corner = new Rect(originX, originY, roomDepth, roomDepth);
            var northWing = new Rect(originX + roomDepth, originY, northCount * roomAlong, roomDepth);
            var westWing = new Rect(originX, originY + roomDepth, roomDepth, westCount * roomAlong);
            var walkNorth = new Rect(northWing.X, northWing.Y + northWing.H, northWing.W, walkW);
            float walkWestY = walkNorth.Y + walkNorth.H;
            var walkWest = new Rect(
                westWing.X + westWing.W,
                walkWestY,
                walkW,
                westWing.Y + westWing.H - walkWestY);
            var lobbyRect = new Rect(originX - lobbySide, westWing.Y + westWing.H, lobbyW, lobbyH);
            var parking = new Rect(
                walkNorth.X + walkW,
                walkNorth.Y + walkNorth.H,
                northWing.X + northWing.W - (walkNorth.X + walkW),
                lobbyRect.Y - (walkNorth.Y + walkNorth.H));
            // Pull-up canopy over the lobby south door — match lobby width so
            // the roof and pillars stay off the east drive / highway.
            var canopy = new Rect(lobbyRect.X, lobbyRect.Y + lobbyRect.H, lobbyRect.W, canopyH);
            var driveSouth = new Rect(
                lobbyRect.X + lobbyRect.W,
                lobbyRect.Y,
                northWing.X + northWing.W - (lobbyRect.X + lobbyRect.W),
                lobbyRect.H + canopyH);

            float west = System.Math.Min(originX, lobbyRect.X);
            float east = northWing.X + northWing.W;
            float south = canopy.Y + canopy.H;
            var bounds = new Rect(west, originY, Down(east - west), Down(south - originY));

            var floor = new BuiltFloor
            {
                Id = spec.Id,
                Name = spec.Name,
                Level = spec.Level,
                Tile = tile,
                Spec = spec,
                Bounds = bounds,
                Content = lobbyRect,
                RoomsPerRow = westCount,
                RoomSize = new Point(roomAlong, roomDepth),
                OutdoorCourt = true,
                WalkWest = walkWest,
                WalkNorth = walkNorth,
                PorteCochere = canopy,
                DriveSouth = driveSouth,
                CornerMass = corner,
                Lobby = lobbyRect,
                Parking = parking
            };

            AddOpenArea(floor, "walk-north", AreaKind.Walkway, "Covered walk", walkNorth);
            AddOpenArea(floor, "walk-west", AreaKind.Walkway, "Covered walk", walkWest);
            AddOpenArea(floor, "walk-north-up", AreaKind.Walkway, "Upper walk", walkNorth, 1);
            AddOpenArea(floor, "walk-west-up", AreaKind.Walkway, "Upper walk", walkWest, 1);
            AddOpenArea(floor, "parking", AreaKind.Parking, "Parking lot", parking);
            AddOpenArea(floor, "parking-drive", AreaKind.Parking, "Drive", driveSouth);
            AddOpenArea(floor, "porte-cochere", AreaKind.Walkway, "Porte-cochère", canopy);

            // Long west wing, rooms nearest the office first, doors facing the lot.
            for (int i = 0; i < westCount; i++)
            {
                var roomRect = new Rect(
                    westWing.X,
                    westWing.Y + westWing.H - (i + 1) * roomAlong,
                    roomDepth,
                    roomAlong);
                AddGuestRoom(floor, spec, roomRect, "east", tile);
            }

            for (int i = 0; i < northCount; i++)
            {
                var roomRect = new Rect(northWing.X + i * roomAlong, northWing.Y, roomAlong, roomDepth);
                AddGuestRoom(floor, spec, roomRect, "south", tile);
            }

            int groundRooms = floor.Rooms.Count;
            for (int i = 0; i < groundRooms; i++)
            {
                var ground = floor.Rooms[i];
                AddGuestRoom(floor, spec, ground.Rect, ground.DoorSide, tile, 1);
            }

            // West half of the walk is the stair tower; east half stays a
            // ground-level lobby door so going to a first-floor room does not
            // force a climb.
            var lobbyStairGap = MakeDoor(lobbyRect, "north", spec.DoorWidth, 0.70f, tile);
            var northDoor = MakeDoor(lobbyRect, "north", spec.DoorWidth, 0.85f, tile);
            var southDoor = MakeDoor(lobbyRect, "south", spec.DoorWidth * 2f, 0.5f, tile);
            // Wide east opening centered on drive traffic so radius-12
            // guests are not sent along the solid wall to the north courtyard door.
            var eastDoor = MakeDoor(lobbyRect, "east", spec.DoorWidth * 2f, 0.65f, tile);
            floor.Areas.Add(new FloorArea
            {
                Id = "lobby",
                Kind = AreaKind.Lobby,
                Label = "Lobby",
                Rect = lobbyRect,
                Walls = true,
                Doors = new List<Door> { lobbyStairGap, northDoor, southDoor, eastDoor }
            });
            floor.LotEntrance = OutsidePoint(northDoor, tile);
            floor.LotEntranceEast = OutsidePoint(eastDoor, tile);
            floor.LotEntranceSouth = OutsidePoint(southDoor, tile);

            float stairH = Down(80);
            float stairFlightW = Down(System.Math.Min(50f, walkW * 0.5f));
            var upperStairs = new Rect(walkWest.X, lobbyRect.Y, stairFlightW, stairH);
            var upDoorSouth = MakeDoor(upperStairs, "south", spec.DoorWidth, 0.5f, tile);
            var upDoorNorth = MakeDoor(upperStairs, "north", spec.DoorWidth, 0.5f, tile);
            floor.Areas.Add(new FloorArea
            {
                Id = "stairs-up",
                Kind = AreaKind.Stairs,
                Label = "Upper stairs",
                Rect = upperStairs,
                Walls = true,
                Doors = new List<Door> { upDoorSouth, upDoorNorth }
            });
            floor.Areas.Add(new FloorArea
            {
                Id = "stairs-up-landing",
                Kind = AreaKind.Stairs,
                Label = "Upper stairs",
                Rect = upperStairs,
                Walls = true,
                Doors = new List<Door> { upDoorSouth, upDoorNorth },
                Level = 1
            });
            floor.UpperStairs = upperStairs;

            float officeH = Down(90);
            float stairW = Down(80);
            float officeLeft = lobbyRect.X + Down(16);
            float officeW = Down(walkWest.X - tile - officeLeft);
            var backBlock = new Rect(
                officeLeft,
                lobbyRect.Y + Down(16),
                officeW,
                officeH);
            var stairsRect = new Rect(backBlock.X, backBlock.Y, stairW, officeH);
            var officeRect = new Rect(
                backBlock.X + stairW,
                backBlock.Y,
                officeW - stairW,
                officeH);
            floor.Areas.Add(new FloorArea
            {
                Id = "stairs",
                Kind = AreaKind.Stairs,
                Label = "Basement stairs",
                Rect = stairsRect,
                Walls = true,
                Doors = new List<Door> { MakeDoor(stairsRect, "east", spec.DoorWidth, 0.5f, tile) }
            });
            floor.Stairs = stairsRect;

            float deskH = Down(40);
            float staffAlley = Down(50);
            floor.FrontDesk = new DeskSpot
            {
                X = backBlock.X + backBlock.W / 2f,
                Y = backBlock.Y + backBlock.H + staffAlley + deskH / 2f,
                W = Down(160),
                H = deskH
            };

            float officeDoorAlong = Geometry.Clamp(
                (floor.FrontDesk.X - officeRect.X) / officeRect.W,
                0.2f,
                0.8f);
            var officeDoor = MakeDoor(officeRect, "south", spec.DoorWidth, officeDoorAlong, tile);
            var officeWest = MakeDoor(officeRect, "west", spec.DoorWidth, 0.5f, tile);
            floor.Areas.Add(new FloorArea
            {
                Id = "office",
                Kind = AreaKind.Office,
                Token = "office",
                Label = "Office",
                Rect = officeRect,
                Walls = true,
                Doors = new List<Door> { officeDoor, officeWest }
            });
            var officeCenter = officeRect.Center;
            floor.Office = new OfficeSpot
            {
                X = officeCenter.X,
                Y = officeCenter.Y,
                W = officeRect.W,
                H = officeRect.H,
                Rect = officeRect,
                Door = officeDoor.Center,
                Approach = OutsidePoint(officeDoor, tile)
            };

            AttachBasement(floor, spec.DoorWidth, tile);
            PruneUnusableDoors(floor.Areas, tile);
            return floor;
        }

        public static BuiltFloor CreateFloor(FloorSpec spec, Rect bounds)
        {
            _ = bounds;
            return CreateMotorCourt(spec);
        }

        public static Rect InnBuildingRect(
            FloorSpec spec,
            float lotWidth,
            float lotHeight,
            float marginX = 48f,
            float marginY = 40f,
            float parkingDepth = 280f)
        {
            _ = lotWidth;
            _ = lotHeight;
            _ = marginX;
            _ = marginY;
            _ = parkingDepth;
            return CreateMotorCourt(spec).Bounds;
        }

        public static Rect AttachParking(BuiltFloor floor, Rect building, float lotHeight)
        {
            _ = building;
            _ = lotHeight;
            return floor.Parking;
        }

        static void AddOpenArea(BuiltFloor floor, string id, string kind, string label, Rect rect, int level = 0)
        {
            if (rect.W <= 0f || rect.H <= 0f) return;
            floor.Areas.Add(new FloorArea
            {
                Id = id,
                Kind = kind,
                Label = label,
                Rect = rect,
                Level = level
            });
        }

        static void AddGuestRoom(BuiltFloor floor, FloorSpec spec, Rect roomRect, string doorSide, float tile, int level = 0)
        {
            int id = floor.Rooms.Count + 1;
            var door = MakeDoor(roomRect, doorSide, spec.DoorWidth, 0.5f, tile);
            floor.Areas.Add(new FloorArea
            {
                Id = $"room-{id}",
                Kind = AreaKind.GuestRoom,
                Token = $"room:{id}",
                Label = $"Room {id}",
                Rect = roomRect,
                Walls = true,
                Doors = new List<Door> { door },
                RoomId = id,
                Level = level
            });
            floor.Rooms.Add(new PlannedRoom
            {
                Id = id,
                Rect = roomRect,
                Center = roomRect.Center,
                DoorSide = door.Side,
                Door = door.Center,
                DoorOpening = door,
                Approach = OutsidePoint(door, tile),
                Level = level
            });
        }

        static void AttachBasement(BuiltFloor floor, float doorWidth, float tile)
        {
            var lobby = floor.Lobby;
            if (lobby.W <= 0f || lobby.H <= 0f) return;

            float Down(float value) => (float)(System.Math.Floor(value / tile) * tile);
            floor.Basement = lobby;
            floor.Areas.Add(new FloorArea
            {
                Id = "basement",
                Kind = AreaKind.Basement,
                Label = "Basement",
                Rect = lobby,
                Walls = true,
                Level = -1
            });

            float storeW = Down(140);
            float storeH = Down(70);
            var storeRect = new Rect(
                lobby.X + Down((lobby.W - storeW) / 2f),
                lobby.Y + Down(24),
                storeW,
                storeH);
            floor.Areas.Add(new FloorArea
            {
                Id = "storage",
                Kind = AreaKind.Storage,
                Label = "Storage",
                Rect = storeRect,
                Level = -1
            });

            foreach (var dept in Departments)
            {
                float deptW = Down(System.Math.Min(170, (lobby.W - tile * 8f) / 2f));
                float deptH = Down(System.Math.Min(80, lobby.H * 0.35f));
                var deptRect = new Rect(
                    dept.Side == "left"
                        ? lobby.X + tile * 2f
                        : lobby.X + lobby.W - deptW - tile * 2f,
                    lobby.Y + lobby.H - deptH - tile * 2f,
                    deptW,
                    deptH);
                var door = MakeDoor(deptRect, dept.Side == "left" ? "east" : "west", doorWidth, 0.5f, tile);
                floor.Areas.Add(new FloorArea
                {
                    Id = dept.Id,
                    Kind = AreaKind.Department,
                    Label = dept.Label,
                    Accent = dept.Accent,
                    Rect = deptRect,
                    Walls = true,
                    Doors = new List<Door> { door },
                    DepartmentId = dept.Id,
                    Level = -1
                });
                var center = deptRect.Center;
                floor.Departments[dept.Id] = new DepartmentSpot
                {
                    Id = dept.Id,
                    Label = dept.Label,
                    Accent = dept.Accent,
                    Rect = deptRect,
                    X = center.X,
                    Y = center.Y,
                    W = deptRect.W,
                    H = deptRect.H
                };
            }
        }

        static Door MakeDoor(Rect rect, string side, float width, float along, float tile)
        {
            string normalized = NormalizeSide(side);
            float doorWidth = (float)(System.Math.Floor(width / tile) * tile);
            bool horizontal = normalized == "north" || normalized == "south";
            float span = (horizontal ? rect.W : rect.H) - doorWidth;
            float offset = (float)(System.Math.Floor((span * along) / tile) * tile);

            if (horizontal)
            {
                return new Door
                {
                    Side = normalized,
                    Width = doorWidth,
                    Center = new Point(
                        rect.X + offset + doorWidth / 2f,
                        normalized == "north" ? rect.Y : rect.Y + rect.H),
                    Normal = new Point(0, normalized == "north" ? -1 : 1)
                };
            }

            return new Door
            {
                Side = normalized,
                Width = doorWidth,
                Center = new Point(
                    normalized == "west" ? rect.X : rect.X + rect.W,
                    rect.Y + offset + doorWidth / 2f),
                Normal = new Point(normalized == "west" ? -1 : 1, 0)
            };
        }

        static string NormalizeSide(string side)
        {
            if (side == "n") return "north";
            if (side == "s") return "south";
            if (side == "e") return "east";
            if (side == "w") return "west";
            return side;
        }

        static Point OutsidePoint(Door door, float tile)
        {
            float reach = tile * 2f;
            return new Point(
                door.Center.X + door.Normal.X * reach,
                door.Center.Y + door.Normal.Y * reach);
        }

        static void PruneUnusableDoors(List<FloorArea> areas, float tile)
        {
            foreach (var area in areas)
            {
                if (area.Doors == null || area.Doors.Count == 0) continue;
                area.Doors.RemoveAll(door =>
                {
                    var probe = new Point(
                        door.Center.X + door.Normal.X * tile * 1.5f,
                        door.Center.Y + door.Normal.Y * tile * 1.5f);
                    foreach (var other in areas)
                    {
                        if (other == area) continue;
                        if (IsOpenSpace(other, probe, tile)) return false;
                    }

                    return true;
                });
            }
        }

        static bool IsOpenSpace(FloorArea area, Point point, float tile)
        {
            if (!PublicKinds.Contains(area.Kind) && area.Kind != AreaKind.Office) return false;
            var r = area.Rect;
            float pad = area.Walls ? tile : 0f;
            return r.Contains(point, pad);
        }
    }
}

File: js/floorplan.js
/**
 * Floorplans are data, not hand-placed geometry.
 *
 * A floor is a vertical stack of bands: rows of guest rooms, corridors, the
 * lobby, and a service band for departments. Each area carries its own walls
 * and door openings. js/nav.js bakes those areas into a walkable grid, so room
 * counts, wings, departments and extra floors are all just numbers here —
 * nothing downstream needs a hand-wired waypoint graph.
 *
 * Every dimension is a multiple of the nav tile, so walls and doors land on
 * exact grid cells instead of half-covering them.
 */

import { CONFIG } from "./config.js";

export const AREA = {
  CORRIDOR: "corridor",
  LOBBY: "lobby",
  GUEST_ROOM: "guestRoom",
  OFFICE: "office",
  DEPARTMENT: "department",
};

/**
 * Ground floor of the flagship hotel. Room rows sit against shared corridors
 * (double-loaded), and vertical side corridors run the full height so every
 * band stays connected without special-case links.
 */
export const FLAGSHIP_GROUND = {
  id: "ground",
  name: "Ground floor",
  level: 0,
  /** Navigation grid resolution in pixels. */
  tile: 10,
  /** Gap between the building shell and any usable space. */
  edge: 20,
  /** Minimum width of the vertical corridors on the far left and right. */
  sideCorridor: 50,
  /** Door openings are this wide — comfortably clear of body radii. */
  doorWidth: 40,
  /** Keep the lobby square: six rooms across, not a ten-room bowling alley. */
  maxRoomsPerRow: 6,
  roomSize: { w: CONFIG.roomWidth, h: CONFIG.roomHeight },
  bands: [
    { kind: "rooms", doorSide: "south" },
    { kind: "corridor", height: 40 },
    { kind: "rooms", doorSide: "north" },
    { kind: "corridor", height: 40 },
    { kind: "lobby", height: 280, grow: true },
    { kind: "service", height: 90 },
  ],
};

/** Areas any character may cross without a permit. */
const PUBLIC_KINDS = new Set([AREA.CORRIDOR, AREA.LOBBY, AREA.DEPARTMENT]);

const DEPARTMENTS = [
  { id: "housekeeping", label: "Mary's room", side: "left", accent: "#e8a0bf" },
  { id: "maintenance", label: "Bob's closet", side: "right", accent: "#ffb347" },
];

/** Tight building shell for a capped room row, centered on the lot. */
export function innBuildingRect(spec, lotWidth, lotHeight, marginX = 48, marginY = 40) {
  const tile = spec.tile;
  const down = (value) => Math.floor(value / tile) * tile;
  const roomW = down(spec.roomSize.w);
  const cols = spec.maxRoomsPerRow ?? 6;
  const innW = cols * roomW + down(spec.sideCorridor) * 2 + down(spec.edge) * 2;
  const innH = lotHeight - marginY - 68;
  return {
    x: down((lotWidth - innW) / 2),
    y: marginY,
    w: innW,
    h: down(innH),
  };
}

/**
 * Build one floor from a spec.
 *
 * @param spec   band/room definition, e.g. FLAGSHIP_GROUND
 * @param bounds building shell rect in pixels
 */
export function createFloor(spec, bounds) {
  const tile = spec.tile;
  const down = (value) => Math.floor(value / tile) * tile;

  const edge = down(spec.edge);
  const content = {
    x: bounds.x + edge,
    y: bounds.y + edge,
    w: down(bounds.w - edge * 2),
    h: down(bounds.h - edge * 2),
  };

  const roomSize = {
    w: down(spec.roomSize.w),
    h: down(spec.roomSize.h),
  };

  // Guest rooms sit flush in a centered block; leftover space widens the
  // vertical side corridors. Room count is therefore a function of space.
  const fit = Math.max(
    1,
    Math.floor((content.w - down(spec.sideCorridor) * 2) / roomSize.w)
  );
  const roomsPerRow = spec.maxRoomsPerRow
    ? Math.min(fit, spec.maxRoomsPerRow)
    : fit;
  const roomsBlockW = roomsPerRow * roomSize.w;
  const roomsBlockX = content.x + down((content.w - roomsBlockW) / 2);

  const bandHeights = resolveBandHeights(spec.bands, content.h, roomSize.h, tile);

  const areas = [];
  const rooms = [];
  let lobby = null;
  let office = null;
  let frontDesk = null;
  const departments = {};

  let cursorY = content.y;
  spec.bands.forEach((band, bandIndex) => {
    const height = bandHeights[bandIndex];
    const bandRect = { x: content.x, y: cursorY, w: content.w, h: height };

    if (band.kind === "rooms") {
      for (let col = 0; col < roomsPerRow; col++) {
        const roomRect = {
          x: roomsBlockX + col * roomSize.w,
          y: cursorY,
          w: roomSize.w,
          h: height,
        };
        const id = rooms.length + 1;
        const door = makeDoor(roomRect, band.doorSide, spec.doorWidth, 0.5, tile);
        areas.push({
          id: `room-${id}`,
          kind: AREA.GUEST_ROOM,
          token: `room:${id}`,
          label: `Room ${id}`,
          rect: roomRect,
          walls: true,
          doors: [door],
          roomId: id,
        });
        rooms.push({
          id,
          rect: roomRect,
          center: rectCenter(roomRect),
          doorSide: door.side,
          /** The opening itself, in the wall. */
          door: door.center,
          doorOpening: door,
          /** Public standing spot just outside the door. */
          approach: outsidePoint(door, tile),
        });
      }
      pushSideCorridors(areas, content, bandRect, roomsBlockX, roomsBlockW);
    } else if (band.kind === "lobby") {
      const lobbyRect = {
        x: roomsBlockX,
        y: cursorY,
        w: roomsBlockW,
        h: height,
      };
      lobby = lobbyRect;

      // Reception reads south→north the way a guest walks in from the lot:
      // double doors, waiting chairs, front desk, then the office (PC behind
      // the counter). Passages on both flanks lead to the guest-room corridors.
      areas.push({
        id: "lobby",
        kind: AREA.LOBBY,
        token: null,
        label: "Lobby",
        rect: lobbyRect,
        walls: true,
        doors: [
          makeDoor(lobbyRect, "north", spec.doorWidth, 0.14, tile),
          makeDoor(lobbyRect, "north", spec.doorWidth, 0.86, tile),
          makeDoor(lobbyRect, "south", spec.doorWidth * 2, 0.5, tile),
          makeDoor(lobbyRect, "west", spec.doorWidth, 0.72, tile),
          makeDoor(lobbyRect, "east", spec.doorWidth, 0.72, tile),
        ],
      });

      const officeW = down(300);
      const officeH = down(100);
      const officeRect = {
        x: lobbyRect.x + down((lobbyRect.w - officeW) / 2),
        y: lobbyRect.y + down(20),
        w: officeW,
        h: officeH,
      };
      const officeDoor = makeDoor(officeRect, "south", spec.doorWidth, 0.5, tile);
      areas.push({
        id: "office",
        kind: AREA.OFFICE,
        token: "office",
        label: "Office",
        rect: officeRect,
        walls: true,
        doors: [officeDoor],
      });
      office = {
        ...rectCenter(officeRect),
        w: officeRect.w,
        h: officeRect.h,
        rect: officeRect,
        door: officeDoor.center,
        approach: outsidePoint(officeDoor, tile),
      };

      const deskH = down(40);
      const staffAlley = down(60);
      frontDesk = {
        x: officeRect.x + officeRect.w / 2,
        y: officeRect.y + officeRect.h + staffAlley + deskH / 2,
        w: down(320),
        h: deskH,
      };

      pushSideCorridors(areas, content, bandRect, roomsBlockX, roomsBlockW);
    } else if (band.kind === "service") {
      areas.push({
        id: `service-${bandIndex}`,
        kind: AREA.CORRIDOR,
        token: null,
        label: "Service corridor",
        rect: bandRect,
        walls: false,
        doors: [],
      });
      for (const dept of DEPARTMENTS) {
        const deptW = down(170);
        const deptRect = {
          x:
            dept.side === "left"
              ? bandRect.x + tile
              : bandRect.x + bandRect.w - deptW - tile,
          y: bandRect.y + tile,
          w: deptW,
          h: bandRect.h - tile * 2,
        };
        areas.push({
          id: dept.id,
          kind: AREA.DEPARTMENT,
          token: null,
          label: dept.label,
          accent: dept.accent,
          rect: deptRect,
          walls: false,
          doors: [],
          departmentId: dept.id,
        });
        departments[dept.id] = {
          id: dept.id,
          label: dept.label,
          accent: dept.accent,
          rect: deptRect,
          ...rectCenter(deptRect),
          w: deptRect.w,
          h: deptRect.h,
        };
      }
    } else {
      areas.push({
        id: `corridor-${bandIndex}`,
        kind: AREA.CORRIDOR,
        token: null,
        label: "Corridor",
        rect: bandRect,
        walls: false,
        doors: [],
      });
    }

    cursorY += height;
  });

  pruneUnusableDoors(areas, tile);

  return {
    id: spec.id,
    name: spec.name,
    level: spec.level ?? 0,
    tile,
    spec,
    bounds,
    content,
    areas,
    rooms,
    roomsPerRow,
    roomSize,
    lobby,
    office,
    frontDesk,
    departments,
  };
}

/** Fixed band heights first; bands marked `grow` share what is left over. */
function resolveBandHeights(bands, totalHeight, roomHeight, tile) {
  const heights = bands.map((band) =>
    band.kind === "rooms"
      ? roomHeight
      : Math.floor((band.height ?? tile * 6) / tile) * tile
  );
  const growIndexes = bands
    .map((band, i) => (band.grow ? i : -1))
    .filter((i) => i >= 0);

  const slack = totalHeight - heights.reduce((sum, h) => sum + h, 0);
  if (slack > 0 && growIndexes.length) {
    const share = Math.floor(slack / growIndexes.length / tile) * tile;
    for (const i of growIndexes) heights[i] += share;
  }
  return heights;
}

/** The strips beside a room block / the lobby are vertical circulation. */
function pushSideCorridors(areas, content, bandRect, blockX, blockW) {
  const left = {
    x: content.x,
    y: bandRect.y,
    w: Math.max(0, blockX - content.x),
    h: bandRect.h,
  };
  const right = {
    x: blockX + blockW,
    y: bandRect.y,
    w: Math.max(0, content.x + content.w - (blockX + blockW)),
    h: bandRect.h,
  };
  for (const [side, rect] of [
    ["left", left],
    ["right", right],
  ]) {
    if (rect.w <= 0) continue;
    areas.push({
      id: `side-${side}-${Math.round(bandRect.y)}`,
      kind: AREA.CORRIDOR,
      token: null,
      label: "Corridor",
      rect,
      walls: false,
      doors: [],
    });
  }
}

/**
 * A door opening on one side of a rect. `along` slides it across that side
 * (0 = start, 1 = end); 0.5 centers it. Snapped to the tile grid.
 */
function makeDoor(rect, side, width, along, tile) {
  const normalized = normalizeSide(side);
  const doorWidth = Math.floor(width / tile) * tile;
  const horizontal = normalized === "north" || normalized === "south";
  const span = (horizontal ? rect.w : rect.h) - doorWidth;
  const offset = Math.floor((span * along) / tile) * tile;

  if (horizontal) {
    return {
      side: normalized,
      width: doorWidth,
      center: {
        x: rect.x + offset + doorWidth / 2,
        y: normalized === "north" ? rect.y : rect.y + rect.h,
      },
      normal: { x: 0, y: normalized === "north" ? -1 : 1 },
    };
  }
  return {
    side: normalized,
    width: doorWidth,
    center: {
      x: normalized === "west" ? rect.x : rect.x + rect.w,
      y: rect.y + offset + doorWidth / 2,
    },
    normal: { x: normalized === "west" ? -1 : 1, y: 0 },
  };
}

function normalizeSide(side) {
  if (side === "n") return "north";
  if (side === "s") return "south";
  if (side === "e") return "east";
  if (side === "w") return "west";
  return side;
}

/**
 * Standing spot just outside a door. Kept close to its own wall so two rooms
 * facing the same corridor do not claim the same spot.
 */
function outsidePoint(door, tile) {
  const reach = tile * 2;
  return {
    x: door.center.x + door.normal.x * reach,
    y: door.center.y + door.normal.y * reach,
  };
}

/**
 * Drop any door that would open into a wall or another private room. Bands can
 * be reordered freely without leaving doors that lead nowhere.
 */
function pruneUnusableDoors(areas, tile) {
  for (const area of areas) {
    if (!area.doors?.length) continue;
    area.doors = area.doors.filter((door) => {
      const probe = {
        x: door.center.x + door.normal.x * tile * 1.5,
        y: door.center.y + door.normal.y * tile * 1.5,
      };
      return areas.some(
        (other) => other !== area && isOpenSpace(other, probe, tile)
      );
    });
  }
}

function isOpenSpace(area, point, tile) {
  if (!PUBLIC_KINDS.has(area.kind)) return false;
  const r = area.rect;
  const pad = area.walls ? tile : 0;
  return (
    point.x >= r.x + pad &&
    point.x <= r.x + r.w - pad &&
    point.y >= r.y + pad &&
    point.y <= r.y + r.h - pad
  );
}

export function rectCenter(rect) {
  return { x: rect.x + rect.w / 2, y: rect.y + rect.h / 2 };
}
